hca/regenerate_api.py

```diff
@@ -0,0 +1,372 @@
+                'required': argrequired and not endpoint_params['seen'],
+                'required_for': [],  # see explanation below
+                'description': arg_params.get('description', ""),
+                'type': arg_params.get('type', None),
+                'format': arg_params.get('format', None),
+                'pattern': arg_params.get('pattern', None)
+            })
+        # If there are more args that weren't seen on this base route, mark then not
+        # required because they're not used in every endpoint.
+        elif i >= len(path_args):
+            positional[i]['required'] = False
+
+        # If the arg is required on this endpoint, add this endpoint to the list of
+        # endpoints for which this arg is needed.
+        if i < len(path_args) and indexed_parameters[path_args[i]]['req']:
+            positional[i]['required_for'].append(path)
+
+
+def _label_optional_args_required(path, endpoint_params, indexed_parameters):
+    """
+    For all optional arguments in an endpoint, determine if the argument is required.
+
+    If an argument is not always required then it will be labeled not required, but the
+    endpoints it is required will be marked to assist the user when using the help option.
+    """
+    options = endpoint_params['options']
+
+    # If in another endpoint with the same base route has a required option that's
+    # missing from this endpoint, it's not required.
+    for (param_name, param_data) in options.items():
+        if param_name not in indexed_parameters:
+            param_data['required'] = False
+
+    for (param_name, param_data) in indexed_parameters.items():
+        if param_data['in'] != "path":  # Only non-positional args
+            argrequired = indexed_parameters[param_name]['req']
+            if param_name not in options:
+                arg_params = indexed_parameters[param_name]
+                options[param_name] = {
+                    'array': param_data['array'],
+                    'required': argrequired and not endpoint_params['seen'],
+                    'required_for': [],
+                    'description': arg_params.get('description', ""),
+                    'type': arg_params.get('type', None),
+                    'format': arg_params.get('format', None),
+                    'pattern': arg_params.get('pattern', None),
+                    'metavar': arg_params.get('metavar', None),
+                    'in': arg_params.get('in', None),
+                    'hierarchy': arg_params['hierarchy']
+                }
+            else:
+                options[param_name]['required'] = argrequired and options[param_name]['required']
+
+            # If the arg is required on this endpoint, add this endpoint to the list of
+            # endpoints for which this arg is needed.
+            if argrequired:
+                options[param_name]['required_for'].append(path)
+
+
+def get_parser(test_api_path=None):
```

Comment:
does this name still reflect what this method does (does not look like it to me)?